### PR TITLE
Run tests for all specified targets

### PR DESCRIPTION
**Pull request by Mandragorian**
Currently cargo-miri uses the first target specified in the command line. However, when multiple targets are specified in a `cargo build` invocation, cargo will build for all of them.

Miri should match this behaviour to reduce surprises.

Fixes: #3460

### CODE REVIEW

File: cargo-miri/src/phases.rs [hunk 1]
```diff
@@ -105,149 +105,148 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
...
+        // interpreting things as flags when it sees the first positional argument.
+
+        // Make sure the build target is explicitly set.
+        // This is needed to make the `target.runner` settings do something,
+        // and it later helps us detect which crates are proc-macro/build-script
+        // (host crates) and which crates are needed for the program itself.
         cmd.arg("--target");
-        cmd.arg(host);
-    }
+        cmd.arg(target);
```
Comment (RalfJung) [inline cargo-miri/src/phases.rs:160]: This does not seem right -- won't this pass `--target` twice if there's also a target set in `args`?

CI passes though so maybe cargo just ignores duplicate targets?

**Review by RalfJung — COMMENTED** [review]
Thanks for the PR!

Given that cargo supports multiple tagets itself, why do we even need to have a loop ourselves? I mean, we need a loop for `setup`,  but then IMO we should just do one big cargo invocation with all the `--target`, and let cargo do the looping.

Also, please add a test.

**Comment by saethlin** [comment]
> I mean, we need a loop for `setup`

I think we actually need support in rustc-build-sysroot for passing through the full list of targets to Cargo, not a loop.

**Comment by RalfJung** [comment]
Why should calling `setup` in a loop not work?

**Comment by saethlin** [comment]
Ah, bad phrasing. It will work, but we don't _need_ to use a loop. It would be much faster to let Cargo run all the sysroot builds in parallel.

**Comment by RalfJung** [comment]
Ah, that's what you mean. Yeah makes sense. It also complicates the copying and printing logic, though.

We can leave that as a future possibility though, it doesn't have to happen in this PR.

**Comment by Mandragorian** [comment]
Looking into the code more, I see that if I don't loop I get an error that `error: only one `--target` argument is supported`, which I think originates from `cargo` and not `miri`. 

I also realised that the loop as I had set it up wouldn't work since `exec(cmd)` never returns. Multiple targets are only supported for the `build` subcommand. So I think that if we want to support running/testing for all targets specified in miri, we might need to spawn the command in a loop afte rall.

**Comment by RalfJung** [comment]
> Multiple targets are only supported for the build subcommand. 

Wait, really? We intend to mirror `cargo test`. If `cargo test` does not support multiple targets then I don't think we should, either.

**Comment by Mandragorian** [comment]
ah my bad. I actually run `cargo miri run` instead of `test`. sorry for the confusion.

File: cargo-miri/src/phases.rs [hunk 2]
```diff
@@ -105,149 +105,148 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
...
+            if arg == "--many-seeds" {
+                many_seeds = Some(DEFAULT_MANY_SEEDS.to_owned());
+            } else if let Some(val) = arg.strip_prefix("--many-seeds=") {
+                many_seeds = Some(val.to_owned());
+            } else {
+                cmd.arg(arg);
+            }
+        }
+        // Forward all further arguments after `--` (not consumed by `ArgSplitFlagValue`) to cargo.
+        cmd.args(args);
```
[the remark this answers is not in the document]
Comment (Mandragorian) [reply cargo-miri/src/phases.rs]: Since this is no longer in the loop the first issue should be addressed.

I added a test that passes 2 targets to miri and expects to see 2 executions of the tests. Is this what you had in mind?

I find the way I did it not so elegant. I am not sure if there is a better way.

If the `cargo_miri` method in the test suite didn't always forward the `--target` flag from its command line to miri, the code might look a bit less complex. But I am not sure if this change is wanted, or if it should be included in this PR.

File: cargo-miri/src/phases.rs (hunk 1 above)
Comment (Mandragorian) [reply cargo-miri/src/phases.rs]: I changed the code to only pass `--target` if it was not specified at all.

File: cargo-miri/src/phases.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply cargo-miri/src/phases.rs]: We must always set `--target`, things will break horribly if we invoke cargo without `--target`.

[the remark this answers is not in the document]
Comment (Mandragorian) [reply cargo-miri/src/phases.rs]: Sorry, let me rephrase. We should set `--target`, yes. I wanted to say that the `cargo_miri` function should not always pass the value of `ARGS.target` to pass the `--target` flag. Maybe it would be simpler if `cargo_miri` accepted the targets as an argument.

Then we could specify exactly which target(s) we want, without having to do all the manipulation I do to ensure that exactly two targets are passed to miri in the tests.

Perhaps we can add a `targets` argument with a default None value. If the value is none, only then we try to use `ARGS.target`, otherwise we use the explicitly provided argument.

File: cargo-miri/src/phases.rs [hunk 3]
```diff
@@ -104,9 +104,14 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
             miri_for_host()
         )
     });
-    let host = &rustc_version.host;
-    let target = get_arg_flag_value("--target");
-    let target = target.as_ref().unwrap_or(host);
+    let mut targets = get_arg_flag_values("--target").collect::<Vec<_>>();
+    let explicit_target = if targets.is_empty() {
```
Comment (RalfJung) [inline cargo-miri/src/phases.rs:108]:
```suggestion
    // If `targets` is empty, we need to add a `--target $HOST` flag ourselves, and also ensure
    // that the host target is indeed setup.
    let target_flag = if targets.is_empty() {
```

File: cargo-miri/src/phases.rs [hunk 4]
```diff
@@ -104,9 +104,14 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
...
+        let host = &rustc_version.host;
+        targets.push(host.clone());
+        Some(host)
+    } else {
+        None
```
Comment (RalfJung) [inline cargo-miri/src/phases.rs:113]:
```suggestion
        // We don't need to add a `--target` flag, we just forward the user's flags.
        None
```

File: cargo-miri/src/phases.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply cargo-miri/src/phases.rs]: Ah, you are talking about run_test.py, not cargo-miri. That context was missing. :)

Yes, please make cargo_miri take a "targets" parameter that defaults to `None`, in which case it uses ARGS.target.

**Review by RalfJung — COMMENTED** [review]
Thanks! Rust side looks good, just some nits.
Test still needs a bit of work.

@rustbot author

File: cargo-miri/src/phases.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mandragorian) [reply cargo-miri/src/phases.rs]: yes, i was talking about `run_test.py`.

I changed the code as discussed. Let me know if that works!

File: test-cargo-miri/run-test.py [hunk 5]
```diff
@@ -180,6 +185,11 @@ def test_cargo_miri_test():
         default_ref, "test.stderr-empty.ref",
         env={'MIRIFLAGS': "-Zmiri-permissive-provenance"},
     )
+    test("`cargo miri test` (multiple targets)",
+        cargo_miri("test", targets = ["aarch64-unknown-linux-gnu", "s390x-unknown-linux-gnu"]),
+        "test.multiple_targets.stdout.ref", "test.stderr-empty.ref",
+        env={'MIRIFLAGS': "-Zmiri-permissive-provenance"},
+    )
```
Comment (RalfJung) [inline test-cargo-miri/run-test.py:192]: This will make `./run-test.py` take quite a bit longer since it needs to build 2 more sysroots... and also on CI we run this many times with different `--target` but this test here is going to always be the same.

Maybe we should have another flag, like `--multi-target`, and only run the test when the flag is set? Then we can set it on CI for the first `run-test.py` invocation and avoid repeating it later.

Comment (Mandragorian) [reply test-cargo-miri/run-test.py]: I added a `--multi-target` flag and made it so that the CI only uses it when testing the host target.

My shell skills are very bad so hopefully the code looks ok, but let me know.

**Comment by RalfJung** [comment]
This looks great, thanks. :)

Can you please squash this into a single commit? (ideally while keeping the same base commit so that the force-push has an empty diff)

**Comment by Mandragorian** [comment]
> This looks great, thanks. :)
> 
> Can you please squash this into a single commit? (ideally while keeping the same base commit so that the force-push has an empty diff)

done!

**Comment by RalfJung** [comment]
Great, thanks!

@bors r+
